aten/src/ATen/native/native_functions.yaml

```diff
@@ -2854,7 +2854,29 @@
 - func: logsumexp.names_out(Tensor self, Dimname[1] dim, bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
   device_check: NoCheck   # TensorIterator
 
+- func: margin_ranking_loss.out(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean, *, Tensor(a!) out) -> Tensor(a!)
+  device_check: NoCheck   # TensorIterator
+  python_module: nn
+  dispatch:
+    CPU, CUDA: margin_ranking_loss_out
+
 - func: margin_ranking_loss(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean) -> Tensor
+  device_check: NoCheck   # TensorIterator
+  python_module: nn
+  dispatch:
+    CPU, CUDA: margin_ranking_loss
```

Comment:
We discussed this and we chose not to go that direction because of the hacky way elementwise + reductions are implemented now in structured kernels. In particular, one needs to do things like `maybe_get_output().resize_({});` and create two different TIs in the most used path. I think structured kernels need a better design when dealing with TI, so I would not be surprised if this pattern was changed in the future, so better to do it "the normal way" for now.

Reply:
Thanks @Lezcano for the explanation! cc @bdhirsh for visibility on structured kernel complications

Reply:
> I think structured kernels need a better design when dealing with TI, so I would not be surprised if this pattern was changed in the future, so better to do it "the normal way" for now.

I agree that those ports are a bit of a hack. But some I think the main reason we ended up porting the other loss functions that way are:

(1) We incur some extra overhead by constructing TI twice, but most loss functions aren't *that* perf critical - that extra overhead seems manageable

(2) The "correct fix" is a pretty complicated undertaking. It would probably complicate `TensorIterator` quite a bit - we'd need a way to introduce "intermediate results" to TI. If that doesn't happen in the near future, then we're left without meta tensor support for all loss functions.

(3) The hack is pretty localized, so if we do end up doing (2) at some point, it doesn't seem like a ton of regretful work to refactor the handle of loss functions afterwards. And then we get to avoid blocking anyone that relies on meta tensor support in the meantime.

cc @ngimel

Reply:
Another way of tackling this would if we added a mechanism (we actually have one) that allowed us to pass the TI from the metakernel to the actual kernel. What would be the problem with that?

Reply:
cc @ysiraichi with whom we discussed these things.
Also, see https://github.com/pytorch/pytorch/issues/70678 which was written after the discussion.